src/forms/TimeInput.jsx

```diff
@@ -19,8 +50,91 @@ class TimeInput extends React.Component {
 	* @description called when the input changes, in turn calls the onChange
 	* 	handler prop, if there is one provided (eg supplied by redux-form or DateTimePicker)
 	*/
-	onChange(e) {
-		this.props.onChange && this.props.onChange(e);
+	onChange(partialState) {
+		if (this.props.onChange) {
+			const stateValues = {
+				...this.state,
+				...partialState
+			};
+
+			const v = `${formatDigits(stateValues.hours) % (this.props.is24Hr ? 24 : 12)}:${formatDigits(stateValues.minutes)}`;
+			this.props.onChange(v);
+		}
+	}
+
+	/**
+	* @function onTimeInputChange
+	* @param e Event Object
+	* @description called when the browser-native time input changes, in turn calls the onChange
+	* 	handler prop, if there is one provided (eg supplied by redux-form or DateTimePicker)
+	*/
+	onTimeInputChange(e) {
+		const { value } = e.target;
+		this.setState(() => ({ value: value }));
+
+		this.props.onChange && this.props.onChange(value);
+	}
+
+	/**
+	* @function onNumberChange
+	* @param e Event Object
+	* @description called when either the hour or minute input changes,
+	* 	in turn changes the state and updates the value of the inputs
+	*/
+	onNumberChange(e) {
+		const { value, id } = e.target;
+
+		this.setState(() => ({ [id]: value }));
+	}
+	/**
+	* @function onMeridianChange
+	* @param e Event Object
+	* @description called when the meridian <select> input changes,
+	* 	in turn changes the state and updates the value of the <select>
+	*/
+	onMeridianChange(e) {
+		const { value } = e.target;
+
+		this.setState(() => ({ meridian: value }));
+	}
+
+	/**
+	* @function onBlur
+	* @param e Event Object
+	* @description called when either the hour or minute input loses focus,
+	* 	and ensures the value entered is not out of the min/max range
+	*/
+	onBlur(e) {
+		const { value, min, max, id } = e.target;
+
+		if (max || min) {
+			const constrainedVal = Math.max(Math.min(value, max), min);
+			if (constrainedVal == value) {
+				this.setState(() => ({ [id]: formatDigits(value) }));
+				this.onChange();
+			} else {
+				this.setState(() => ({ [id]: formatDigits(constrainedVal) })); // try doing onChange stuff in this setstate callback
+				this.onChange({ [id]: formatDigits(constrainedVal) });
+			}
+		}
+	}
+
+	componentDidMount() {
+		this.setState(() => ({ supportsTime: !this.props.forceTextInput && this.inputEl.type === 'time' }));
+	}
+
+	componentWillReceiveProps(nextProps) {
+		if (this.props.value !== nextProps.value) {
+			const props = nextProps;
+			this.setState({
+				hours: props.value && (formatDigits(getTimeParts(props.value, 'hours') % (props.is24Hr ? 24 : 12)) || '12').toString(),
+				minutes: props.value && (formatDigits(getTimeParts(props.value, 'minutes')) || '00').toString(),
+			});
+		}
+	}
+
+	shouldComponentUpdate(nextProps, nextState) {
```

Comment:
I'm forgetting why I put this here. Maybe it was from another version I tried. this may be able to be removed.

Reply:
seems like a premature optimisation to me... if I were you I think I'd chop it out. Also, we'd probably need to do a deep equals comparison here because i'm guessing `this.state` is not the same object reference as `nextState`, right?